### PR TITLE
Proper handling of offset and scale in metadata picking

### PR DESCRIPTION
**Pull request by javagl**
# Description

Addresses https://github.com/CesiumGS/cesium/issues/12236 

The metadata picking did not take into account the `offset/scale` that may be defined in a class property or property texture property. 

The solution is to ensure that the `offset/scale` (and normalization) are properly handled when converting the metadata values into values that are supposed to be written into the metadata picking frame buffer. For example, a normalized `UINT8` property with `offset=1.0` and `scale=2.0` will be converted into `1.0 + 2.0 * ~0.5 = 2.0` by the standard metadata handling, But before writing this into the frame buffer, it has to be converted back into the [0, 1] range (reflecting the [0,255] range that can be stored in the frame buffer). 

After reading the value from the frame buffer, it has to be converted back into the proper metadata value. These operations strongly resemble the [`MetadataComponentType.apply/unapplyValueTransform`](https://github.com/CesiumGS/cesium/blob/baaabaa49058067c855ad050be73a9cdfe9b6ac7/packages/engine/Source/Scene/MetadataComponentType.js#L360), but have additional caveats

- **Every** value must be mapped to the [0, 1] range
- This mapping is not done by "implementing it", but by _generating shader code that implements it_
- An `offset/scale` that is defined by a 'class property' may actually be overridden by the 'property texture property'

The last point may make it necessary to store the "effective" `offset/scale` inside the [`PickedMetadataInfo`](https://github.com/CesiumGS/cesium/blob/baaabaa49058067c855ad050be73a9cdfe9b6ac7/packages/engine/Source/Scene/PickedMetadataInfo.js). Right now, it only stores the `MetadataClassProperty`. But when the `offset/scale` have been defined in the `PropertyTextureProperty` or a `PropertyAttributeProperty` (!), then these `offset/scale` values have to be used for the conversions.

## Issue number and link

https://github.com/CesiumGS/cesium/issues/12236

## Testing plan

Details TBD - started creating specs

# Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [x] I have updated `CHANGES.md` with a short summary of my change
- [x] I have added or updated unit tests to ensure consistent code coverage
- [x] I have updated the inline documentation, and included code examples where relevant
- [x] I have performed a self-review of my code

### CODE REVIEW

**Comment by javagl** [comment]
It's never as easy as it looks...

> The last point may make it necessary to store the "effective" offset/scale inside the [PickedMetadataInfo](https://github.com/CesiumGS/cesium/blob/baaabaa49058067c855ad050be73a9cdfe9b6ac7/packages/engine/Source/Scene/PickedMetadataInfo.js).

This is not possible. At the point where the actual value is obtained (namely, inside the shader), it is not longer possible to know where any `offset/scale` came from to begin with. So it's possible to "undo" the value transform in the shader (with some quirks). And it is possible to write that value in [0,1] into the color (ending up as [0, 255] in the frame buffer). But it is not possible to convert this back to the _right_ value after _reading_ it from the frame buffer, because at that point, we don't know whether the "value transform" that was applied came from a 'class property' or from the 'property texture/attribute property'.

There _might_ be solutions to that, but all approaches that I can think of right now are ... complicated, to say the least. We certainly don't want to replicate the whole logic of value transforms (and things like `KHR_texture_transform` _inside the metadata picking shader function_ ... 🤔

**Comment by javagl** [comment]
The solution to the previous point turned out to be not toooo complicated.

When the picking starts, it will obtain the `StructuralMetadata` from the object, and then either obtain the `PropertyTextureProperty` or the `PropertyAttributeProperty` that corresponds to the picked metadata. (This may require some more thought - e.g. could there be an attribute and a texture that describe the same property...? Just thinking about corner cases and limitations here...). 

This object always offers the `offset/scale`, _in its final form_ - i.e. even when no `offset/scale` have been defined, this will be filled with default values, but also handle possible overrides from the class property. This object is then put into the `PickedMetadataInfo`, and used 

1. for creating the shader code that maps the metadata values to [0, 1] for the picking
2. for converting the value in [0,255] from the frame buffer back into its original range

I started adding the corresponding tests. In that process I stumbled over an issue where the shader compilation bails out for [normalized `UINT8` properties in property textures with offset/scale being defined in the class property](https://github.com/CesiumGS/cesium/pull/12237/commits/ed6ad00c538d19c26524f3478f48794591665d66#diff-e7104646fcc3649a63c2a9e7468b5143778439156f8f4c56fc0e20a8256c4823R3594), and it looks like this is independent of the picking (and therefore may have to be moved into a dedicated issue), but I have to verify that.

I'd also like to do more tests with property _attributes_ (because these are currently not covered at all). Since the picking only relies on the metadata, and this does not even "know where it comes from" (attribute or texture), this _might_ just work transparently. But now there is some code that explicitly has to handle _both_ textures and attributes, so that should be covered by specs as well.

**Comment by javagl** [comment]
So, I _did_ try out some of this with property attributes. Or rather _tried_ to try it. There are more issues with that, and see the handling of property attributes for metadata picking as a part of https://github.com/CesiumGS/cesium/issues/12225 , likely to be spun out into a dedicated issue. (Handling of _arbitrary_ types being moved through the frame buffer _does_ raise quite a few questions).

So I'll consider this as ready for review, given that it contains several spec test cases, and I tried it out with the data from https://github.com/CesiumGS/cesium/pull/12232 (where this originally came up), and there, the current state seems to work as expected.

I did have to re-run some CI steps a few times. We do have some flaky tests there...

**Comment by javagl** [comment]
A small update: As mentioned above, pricking arbitrary property attributes is not supported until now. It is not entirely clear what the expected behavior should be when someone _tries_ to pick property attributes. But in the previous state, it could happen that this caused an error to be thrown (specifically, when the property attribute type was one of the types that could not trivially be transported through the picking frame buffer, like a `FLOAT32` or `ENUM`).

The last commit handles this: The function that determines the "picked metadata property" now **only** looks for property _texture_ properties (and no longer for property _attribute_ properties), and bails out early if no property texture property is found. A `// Note...` was added to the function, pointing to https://github.com/CesiumGS/cesium/issues/12225 , which keeps track of this.

**Comment by ggetz** [comment]
Hi @javagl, I'm taking a look at this PR now. Do you have an updated testing plan that I can use to verify the picked vaues?

File: packages/engine/Source/Scene/MetadataPicking.js
Comment (ggetz) [inline packages/engine/Source/Scene/MetadataPicking.js:267]:
```suggestion
 * Converts the given type into a raw value or array representation.
```

File: packages/engine/Source/Scene/MetadataPicking.js
Comment (ggetz) [inline packages/engine/Source/Scene/MetadataPicking.js:274]: Since this "type" is used repeatedly, I would suggest using [`typedef`](https://jsdoc.app/tags-typedef) to define it fro both convenience, accuracy, and specificity.

For instance, I think that `MetadataPicking.convertToObjectType` has incomplete types for the `value` parameter. 

And `Scene.prototype.pickMetadata` could be updated to use this more specific type instead of `any`.

Something like:
```
/**
 * A raw value or array representation of a metadata value
 * @typedef {(number|string|boolean|number[]|undefined)} MetadataType.packedValue
 */
```

I maybe would suggest doing the same for the instance values, like:
```
/**
 * An instance of a metadata value
 * @typedef {(number|string|boolean|Cartesian2|Cartesian3|Cartesian4|Matrix2|Matrix3|Matrix4|undefined)} MetadataType.instanceValue
 */
```

**Review by ggetz — CHANGES_REQUESTED** [review]

File: packages/engine/Source/Scene/MetadataPicking.js
Comment (javagl) [reply packages/engine/Source/Scene/MetadataPicking.js:302]: The `DeveloperError` seems to be the blanket handling of "Something is wrong". In most cases, it's not the _developer_ who is "responsible" for anything there. In many cases, it's some sort of "data error". I'll just throw a `RuntimeError` here, given the lack of further differentiation.

Comment (ggetz) [reply packages/engine/Source/Scene/MetadataPicking.js:302]: I see your point. A runtime error would be valid, so long as it is properly bubbled up and does not stop the render loop.

File: packages/engine/Source/Scene/MetadataPicking.js
Comment (javagl) [reply packages/engine/Source/Scene/MetadataPicking.js:300]: The `unpack` was a copy-and-paste error, and it did not show up due to the limitations of the types that can appear in property textures in general.

More broadly, this is related to https://github.com/CesiumGS/cesium/pull/12075#discussion_r1781538033 : It's unfortunate that it was necessary to create these back-and-forth conversion functions _here_ . This PR (and the linked one) _should_ have _used_ existing functions (which _should_ already be covered by specs).

But ... ignoring the spilled milk: If course, the current coverage of the `MetadataPicking` is dismal, and I started creating specs for that. Given the combinatorial explosion, this is ongoing, and will at some point have to be aligned with the existing structures, as mentioned in the comment that I linked to above.

File: packages/engine/Source/Scene/PickedMetadataInfo.js
Comment (javagl) [reply packages/engine/Source/Scene/PickedMetadataInfo.js:62]: Related to https://github.com/CesiumGS/3d-tiles/issues/650 ...

If the `hasValueTransform/offset/scale` were defined in some common `TransformableValue` interface, then this would be sufficient. But as it stands, `PropertyTextureProperty` and `PropertyAttributeProperty` are distinct, unrelated classes, and it is not clear whether the consumers of the `metadataProperty` will ever try to access the `textureReader` of the former.

I _think_ that it is important to *know* what the users of this class rely on. But if desired, I can remove that information.

File: packages/engine/Source/Scene/MetadataPicking.js
Comment (javagl) [reply packages/engine/Source/Scene/MetadataPicking.js:274]: This is somehow related to https://github.com/CesiumGS/cesium/pull/12075#discussion_r1777461460 

The type information for metadata is difficult. You mentioned
`(number|string|boolean|Cartesian2|Cartesian3|Cartesian4|Matrix2|Matrix3|Matrix4|undefined)`

But in reality, in many places, it would be something like

`(bigint|number|string|boolean|Cartesian2|Cartesian3|Cartesian4|Matrix2|Matrix3|Matrix4|bigint[]|number[]|string[]|boolean[]|Cartesian2[]|Cartesian3[]|Cartesian4[]|Matrix2[]|Matrix3[]|Matrix4[]|undefined)`

(Give and take a few, depending on the subtle limitations here and there that are _highly_ specific for the context (e.g. for property textures, or implementations that are "incomplete" in terms of types that actually are supported)

It would be unfortunate if this tempted people to not even _trying_ to document things properly, but just use the
```
/**
 * @private
 */
```
that currently appears 349 times in the codebase.

I'd _like_ to be more specific. And I can _try_ to be more specific. But thoughness takes time...

Comment (ggetz) [reply packages/engine/Source/Scene/MetadataPicking.js:274]: No need to fix this _everywhere_ in the code base for this PR. Would it be sensible to create the relevant type definitions with our best list of possible types and use them in the scope of this PR?

We can submit a cleanup issue that can cover the remaining cases.

I'm harping on this because the types are exposed through a public API, and the typings will help users of the library, especially those using typescript.

File: packages/engine/Source/Scene/PickedMetadataInfo.js
Comment (ggetz) [reply packages/engine/Source/Scene/PickedMetadataInfo.js:62]: I agree with the sentiment. Class-design aside, that _should_ the point of providing type information.

File: packages/engine/Source/Scene/MetadataPicking.js
Comment (ggetz) [reply packages/engine/Source/Scene/MetadataPicking.js:300]: Thanks. Some tests are better than no tests!

Comment (javagl) [reply packages/engine/Source/Scene/MetadataPicking.js]: I added (more) specs, and tried to handle the combinatorial explosion by being exhaustive in the "low-level" functions, and picking "representative cases" in the higher-level ones. 

The result is currently https://github.com/CesiumGS/cesium/blob/49bd80e6aea84b6f467e5fca9df6d9c52ebd66e5/packages/engine/Specs/Scene/MetadataPickingSpec.js 

I've seen few cases with nested `describe`s, as in
```
describe("The class...") {
    describe("the function...") {
       it("does this" ...),
       it("does that" ...),
    }
}
```
but thought that this could make sense here (and ... in many other classes, actually...)

If this should be "flattened", just drop me a note.

File: packages/engine/Source/Scene/PickedMetadataInfo.js
Comment (javagl) [reply packages/engine/Source/Scene/PickedMetadataInfo.js:62]: So... I assume that this is resolved...? 
(If the solution is "REMOVE IT!", then you may have to write "REMOVE IT!" - I'm not soooo good at this 😆 )

File: packages/engine/Source/Scene/MetadataPicking.js
Comment (javagl) [reply packages/engine/Source/Scene/MetadataPicking.js:292]: Are you opposed to keeping the `{any}` in this (private!) part for now, and tracking possible refinements of the JSDoc for better TypeScript-consumabiltiy in a dedicated issue?

**Comment by javagl** [comment]
I addressed most of the inlined comments, and added specs.

The main open point seems to be the type handling that was already brought up in https://github.com/CesiumGS/cesium/pull/12075#discussion_r1777461460 . While it would be desirable to have stricter type information, the variety of metadata types makes it difficult to ~"quickly come up with" something, and I'd say that more thought has to go into that. (This is amplified by the fact that in the TypeScript world, these types are part of the API, and every change there is a _breaking_ change). 

In terms of testing: The newly added specs dedicatedly cover the case that there [is an offset and scale](https://github.com/CesiumGS/cesium/blob/49bd80e6aea84b6f467e5fca9df6d9c52ebd66e5/packages/engine/Specs/Scene/MetadataPickingSpec.js#L1422). 

I can try to create further samples, and it would actually be good to extend the existing `3d-tiles-samples` with cases that use offset and scale (currently, they do not cover this at all). But the schedule for doing this within this PR may be a bit tight...

File: packages/engine/Source/Scene/MetadataPicking.js
Comment (ggetz) [reply packages/engine/Source/Scene/MetadataPicking.js]: Nested defines are all good.

File: packages/engine/Source/Scene/PickedMetadataInfo.js
Comment (ggetz) [reply packages/engine/Source/Scene/PickedMetadataInfo.js:62]: Sorry for the ambiguity: REMOVE IT! 😆

**Comment by ggetz** [comment]
> This is amplified by the fact that in the TypeScript world, these types are part of the API, and every change there is a breaking change

Narrowing the types may be a breaking change, but I don't think adding possible types should be breaking. LMK if you disagree.

**Comment by ggetz** [comment]
> The build is failing at "verify batched features statistics". I've probably restarted this a dozen times already, and 3 times in this PR. We should disable this test.

Unfortunately, the test failures due to https://github.com/CesiumGS/cesium/issues/11958 are a race condition. It is not the failing test that is itself the problem, so the error will just happen in another test if we ignore that one.

We need to spend some time tracking that race condition down soon.

File: packages/engine/Source/Scene/PickedMetadataInfo.js
Comment (javagl) [reply packages/engine/Source/Scene/PickedMetadataInfo.js]: Unresolving: 

I stumbled over this occasionally. For the suggested change and the resulting build error at https://github.com/CesiumGS/cesium/actions/runs/11584611448/job/32252079758?pr=12237#step:6:2897 , there are basically the two-and-a-half options:

- Refer to the type without including it (causing the current error)
- Include the type (which causes a linting error)
- Do not refer to the type - which defeats the purpose of the type information

@ggetz  Any pattern of how to solve this?

Comment (ggetz) [reply packages/engine/Source/Scene/PickedMetadataInfo.js]: I think this is occurring because `PropertyTextureProperty` are `PropertyAttributeProperty` are marked as private. (They are also both marked as experimental, and that isn't necessary if the API is private.) Otherwise, there should be no need to import the class for the typing to work.

When we take a pass on the metadata typing, we should probably check to make sure private classes are appropriately marked as such or not.

Ok to skip for now.

**Comment by ggetz** [comment]
Hi @javagl, I wanted to check on the status of this PR. Are there any other updates planned, or is this waiting on my final review?

**Comment by javagl** [comment]
There are two points that still have to be addressed:

- The build failure, related to JSDoc referring to the (private) `PropertyTextureProperty/Attribute`. I'd fix this by not "hot"linking them
- The `{any}` return type, summarized in https://github.com/CesiumGS/cesium/issues/12269 

I'd still like to address the latter as well. And it's one of the next points on my TODO list. This typing _might_ not make its path into this PR (depending on how well this works), but I'd at least give it a stab. (Even if we cannot be more specific than `any` here, this PR should become part of the next release)

**Comment by javagl** [comment]
@ggetz A preview for how the type could be made more specific than `{any}` is in https://github.com/CesiumGS/cesium/pull/12315 

If you agree, I'd merge that draft PR state into _this_ one.

**Comment by ggetz** [comment]
Thanks for the updates here @javagl! 

This is looking good to me based on all of of discussion above, including the newer type updates.

**Review by ggetz — APPROVED** [review]

**Pull request merged**
